drivers: dma: Introduce driver for NXP's eDMA IP

LaurentiuM1234 opened the pull request:
Please check commit c63a3e7107c1977b6c665000f9e420013cec37d5's description for reason why this driver is used instead of `dma_mcux_edma`.

LaurentiuM1234 committed:
drivers: dma: Introduce driver for NXP's eDMA IP

This commit introduces a driver for NXP's eDMA IP.

The main reasons for introducing a new driver are the following:

	1) The HAL EDMA wrappers don't support well different
	eDMA versions (e.g: i.MX93 and i.MX8QM). As such, a new
	revision had to be introduced, thus requiring a new Zephyr
	driver.

	2) The eDMA versions found on i.MX93, i.MX8QM, and i.MX8QXP
	don't use the DMAMUX IP (instead, channel MUX-ing is performed
	through an eDMA register in the case of i.MX93).

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

LaurentiuM1234 committed:
manifest: Bump up hal_nxp revision

Bump up hal_nxp revision to contain the EDMA REV2-related
changes from NXP HAL.

Signed-off-by: Laurentiu Mihalcea <laurentiu.mihalcea@nxp.com>

teburd commented:
I’m ok with having another eDMA, what I did with the designware dma is try to create a common set of functions and reuse them where possible as there are many users of the designware gpdma. It would be nice if eDMA did something similar if possible. Like if there’s common transfer block definitions and configuration registers for example some helpers could be made to reuse between the two.

LaurentiuM1234 commented:
> I’m ok with having another eDMA, what I did with the designware dma is try to create a common set of functions and reuse them where possible as there are many users of the designware gpdma. It would be nice if eDMA did something similar if possible. Like if there’s common transfer block definitions and configuration registers for example some helpers could be made to reuse between the two.

ACK. If everyone's ok with this, I'd like to have this in first as-is before making such clean-ups. It would make it a bit easier for me and I'm thinking it would also ease the review process a bit.

DerekSnell reviewed (changes requested):
Hi @LaurentiuM1234 ,
Sorry for interfering, but I am opposed to naming drivers and bindings after NXP family names, and I am trying to stop NXP from doing this with new drivers we upstream.  NXP re-uses the same hardware IP across different families, and we already have confusion where a binding named for one family is used on another.  I am also opposed to calling this new driver the `imx_edma` because the i.MX RT family of MCUs is branded under the i.MX name, also uses the eDMA which is already upstreamed, but apparently a different eDMA flavor than this one.

As we have been naming new drivers/bindings, we are trying to use the NXP IP name to differentiate.  Those names are not always obvious from the public documentation.  But I think if we work together internally, we can come up with a better name than `imx_edma` for this.  Do you already know our name for this IP, and how NXP tells it apart from the existing upstream eDMA?

Thanks

LaurentiuM1234 commented:
> Sorry for interfering, but I am opposed to naming drivers and bindings after NXP family names, and I am trying to stop NXP from doing this with new drivers we upstream. NXP re-uses the same hardware IP across different families, and we already have confusion where a binding named for one family is used on another. I am also opposed to calling this new driver the `imx_edma` because the i.MX RT family of MCUs is branded under the i.MX name, also uses the eDMA which is already upstreamed, but apparently a different eDMA flavor than this one.

I agree the name isn't the best so this is pretty reasonable.



> But I think if we work together internally, we can come up with a better name than `imx_edma` for this. Do you already know our name for this IP, and how NXP tells it apart from the existing upstream eDMA?

Mm, not sure actually. Took at look at i.MX RT1050's TRM and there seem to be a couple of differences:

1. RT has DMAMUX
2. RT has somewhat different register layout (no MP region, just TCD + of course, some registers differ)

Also took a look at how this is handled in Linux: seems like RT uses `fsl-edma.c`, while 93 for example uses `fsl-edma-v3.c`. As such, I propose 2 naming schemes:

1. If there's ever a scenario in which MPUs and MCUs use the same IP name but the IP works somewhat differently we go for `{subsys}_mcux_{ip_name}` for MCUs (e.g: dma_mcux_edma) and `{subsys}_{ip_name}` for MPUs (e.g: dma_edma).
2. We do the same thing as Linux: add a version to the name. In this case, we'd have `dma_edma_v2.c`. We'd have to specify in the binding how the IP is to be used (e.g: we use `dma_edma_v2` for the EDMA IPs found on boards such as i.MX93, i.MX8QM etc...)

BTW, even boards such as i.MX93EVK may include different eDMA versions (e.g: see eDMA3 and eDMA4 which have slightly different register layouts and channel numbers) so I'm not sure how we can tell the difference.

danieldegrasse commented:
Hi @LaurentiuM1234, as you mention it does seem that iMX93 has two EDMA controller IP revisions: EDMAv3 and EDMAv4. As far as EDMAv4, I am not sure we have support for this yet in Zephyr, and I'm not familiar enough with the IP to know if we need another driver for it.

However, EDMAv3 (at least with DMAMUX) was already supported by #61311. Could we consider decoupling the edma driver and the DMAMUX code somehow? Or compiling it out based on the driver compatible? I would like to reuse code as much as possible here when the underlying IP is similar.

LaurentiuM1234 commented:
> Hi @LaurentiuM1234, as you mention it does seem that iMX93 has two EDMA controller IP revisions: EDMAv3 and EDMAv4. As far as EDMAv4, I am not sure we have support for this yet in Zephyr, and I'm not familiar enough with the IP to know if we need another driver for it.

This driver supports both of 93's versions.



> However, EDMAv3 (at least with DMAMUX) was already supported by #61311. Could we consider decoupling the edma driver and the DMAMUX code somehow? Or compiling it out based on the driver compatible? I would like to reuse code as much as possible here when the underlying IP is similar.

What about the HAL revision? This driver uses a different NXP HAL EDMA revision. The point of this new revision is to provide a common API to the upper layers that can be used for any EDMA revision (or at least this is the plan, for now it works OK with EDMA3 and EDMA4). I've already had a lot of trouble dealing with the fact that EDMA4 from i.MX93 uses a different DMA_Type structure so I'm not willing to go back to the old revision and "somehow" make it work for all EDMA versions.

LaurentiuM1234 commented on drivers/dma/dma_nxp_imx_edma.h:
```diff
@@ -0,0 +1,478 @@
+		}
+		break;
+	case CHAN_STATE_STOPPED:
+		if (next != CHAN_STATE_CONFIGURED) {
```
I don't think I see a reason for which we'd allow this transition. The `INIT` state is used to mark the fact that the channels are, well, initialized. It shouldn't be mistaken with the `CONFIGURED` state which means you've committed a configuration to the hardware.

I guess `INIT` can also signify that either the hardware is in an initial known state (registers are 0 and what not)(this is not supported ATM) or is in an unknown state.

LaurentiuM1234 commented on drivers/dma/dma_nxp_imx_edma.h:
```diff
@@ -0,0 +1,478 @@
+		}
+		break;
+	case CHAN_STATE_CONFIGURED:
+		if (next != CHAN_STATE_CONFIGURED &&
```
I guess logically speaking there's nothing wrong with re-configuring an already configured channel. Not sure why you'd do this though.

For now, we're going to forbid this transition (assuming SOF doesn't call config() twice as it does for the DAI, which I'll have to test) because `edma_config()` will fail if called a second time while the channel state is != `STOPPED`. This is because the function setting the MUX value queries the current MUX value and if != 0 it will fail. The only function clearing the MUX value is `edma_stop()`, thus the need for the channel to be in `STOPPED` state for the second `edma_config()` call.

If we ever find a case in which `edma_config()` is called twice we can allow the transition and fix the MUX setting issue pretty easily.

LaurentiuM1234 commented:
**V2 updates**

1. Fixed some typos
2. Added comments to `edma_chan_cyclic_produce()` and `edma_chan_cyclic_consume()`.
3. Added comments to `struct edma_channel`'s fields.
4. CONFIGURED->CONFIGURED transition is no longer allowed.
5. `EDMA_CHAN_PRODUCE_CONSUME_B()` call inside `edma_reload()` is now locked. This is because both the ISR and `edma_reload()` update the `free` and `pending_length` which may lead to a race condition.

LaurentiuM1234 commented:
**V3 updates**

- Some naming-related changes to align with the new version of the NXP HAL EDMA driver. No functional changes.

LaurentiuM1234 commented:
**V4 updates**

1. Changed naming to `dma_nxp_edma`. No functional change.

LaurentiuM1234 commented:
@DerekSnell could you please review this again? Thanks!

decsny reviewed (dismissed):
I do not think it is efficicient to maintain 2 EDMA drivers if it's not necessary. Can you please be more concrete about why we cannot reuse code for the legacy EDMA versions, because I do not understand from your comments.

DerekSnell reviewed (dismissed):
Hi @LaurentiuM1234,

Thank you for updating this PR.  There are still some references to i.MX I want removed, since this driver could be used on other SOCs.  But once you remove those, I will remove my block.  Thank you

LaurentiuM1234 commented:
> Hi @LaurentiuM1234,
> 
> Thank you for updating this PR. There are still some references to i.MX I want removed, since this driver could be used on other SOCs. But once you remove those, I will remove my block. Thank you

Should be fixed now

iuliana-prodan commented on drivers/dma/dma_nxp_edma.c:
I see that here you return in case of error.
On EDMA_CHAN_PRODUCE_CONSUME_A() you don't.

LaurentiuM1234 replied:
As opposed to the ISR, here we can actually let the user know something went wrong. There's also no callback here which should (arguably) be called even if something bad happened.

LaurentiuM1234 commented on drivers/dma/dma_nxp_edma.c:
> It can recover from this?
> I see you don't return from this error.

At the moment no, it can't since we don't have the infrastructure for this (i.e: there's no way for the upper layers to tell if this happened).


> IIRC this will fail at some point - it throws overflow/underrun a few times and then it crashes.

I don't recall SOF crashing from this. Most of the times the crash happened because of the code from the SAI driver.

> If not, then we should return directly.

I think this depends. The transfer did indeed happen successfully so in my opinion it wouldn't be wrong to notify the upper layer through the callback but then again I don't have an actual use case for this (we don't use the callback for SOF) so I'm not 100% sure.

iuliana-prodan replied:
Ok, so actually `edma_isr()` is not used now.
Let's leave it as is and, when using it, we'll update it if it's the case.

LaurentiuM1234 commented on drivers/dma/dma_nxp_edma.c:
Fixed

LaurentiuM1234 commented:
dependency merged, should be good to go as well once CI is green

dbaluta commented:
@teburd @dleach02 good to merge. dependency merged and SHA updated.
